src/client/components/dimension-list-tile/dimensions-renderer.tsx

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2017-2018 Allegro.pl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from "react";
+import { SearchableFolder } from "../searchable-tile/searchable-folder";
+import { DimensionClickHandler, DimensionItem } from "./dimension-item";
+import { DimensionGroupForView, DimensionOrGroupForView, DimensionForView, DimensionForViewType } from "./dimensions-converter";
+
+export class DimensionsRenderer {
```

Comment:
No need for using classes here - you're not containing any sort of state.
Consider using multiple named exports, and importing them as regular functions - FP style. ;)

Reply:
We'll stick with the class here.
We don't want to extend functions' signatures to include additional arguments and the class indeed has a state - it encapsulates handlers and a search text which are being used to create components.

Reply:
You can always refactor this to use higher order functions ;)

Reply:
We could of course refactor it to a function returning another function and use some variables from a surrounding closure in inner function.
I am probably JVM biased ;-) but I seem to like it better structured this way.

When it comes to multiple instantiations in render() I suppose that its one call to a constructor function + short lived reference allocation (probably even in stack only thanks to escape analysis). When it starts expressing itself as a performance problem we will for sure get back to it and refactor.

Reply:
Well, I'm certainly not gonna impose any sort of architectural direction here. Just trying to call to your attention that this is in fact react antipattern. `render` method is called multiple times throughout lifecycle of the page - it is used to calculate desired component tree on every possible (for given piece of tree) state change.
Be sure to pay special attention to this while testing.